capstone/src/main/java/com/google/sps/servlets/BlackOwnedRestaurantsDataServlet.java

```diff
@@ -14,82 +14,107 @@
 
 package com.google.sps.servlets;
 
+import com.google.maps.model.PlaceDetails;
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
 import com.google.gson.Gson;
 import com.google.sps.data.RestaurantDetailsGetter;
+import com.google.sps.data.RestaurantQueryHelper;
+import com.google.sps.data.Restaurant;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.List;
 import java.util.HashSet;
-import java.util.Scanner;
-import com.google.maps.model.PlaceDetails;
+import java.util.Set;
+import java.util.Arrays;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.FetchOptions.Builder;
+import java.util.Scanner;
 
 /** Returns black owned restaurants data as a JSON object */
-@WebServlet("/black-owned-restaurants-data")
+@WebServlet("/black-owned-restaurants")
 public class BlackOwnedRestaurantsDataServlet extends HttpServlet {
+  private static final String DATABASE_NAME = "BlackOwnedRestuarants";
 
-  private ArrayList<String> blackRestaurants = new ArrayList<>();
-  private ArrayList<PlaceDetails> detailedPlaces = new ArrayList<>();
-  
   private RestaurantDetailsGetter details = new RestaurantDetailsGetter();
+  private RestaurantQueryHelper queryHelper = new RestaurantQueryHelper();
+
+/** scrapes business names from source */
+  private ArrayList<String> getRestaurantNames() throws IOException{
+    ArrayList<String> blackOwnedRestaurants = new ArrayList<>();
 
-  @Override
-  public void init() {
     Scanner scanner =
         new Scanner(
             getServletContext().getResourceAsStream("/WEB-INF/black-owned-restaurants.csv"));
-    int i = 0;
+
     while (scanner.hasNextLine()) {
-      if (i >= 1) break;
       String line = scanner.nextLine();
       String[] cells = line.split(",");
 
       String name = String.valueOf(cells[0]);
-      detailedPlaces.add(details.request(name));
 
-      blackRestaurants.add(name);
-      i++;
+      blackOwnedRestaurants.add(name);
     }
     scanner.close();
+    return blackOwnedRestaurants;
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String keywords = (String) request.getParameter("keyword");
+    Query query = new Query(DATABASE_NAME).addSort("rating", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Entity> allRestaurants = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
```

Comment:
It looks like we're pulling out all of the restaurants in the database, and then checking them one by one to see if they have the keywords?

Ideally, we'd be using a Filter https://cloud.google.com/datastore/docs/concepts/queries#datastore-datastore-property-filter-java to only load the restaurants that match the filter.

If that's infeasible, we should chat about alternatives.

Reply:
I looked into filters before and it seemed like it could only filter based on comparison operators like =, >, <
An alternative I found was using GQL, which allows you to write more sql like commands, but I had some trouble implementing it to work with an appengine app

Reply:
oh I just reread the Filter description more closely and it seems like if the property is array-valued, the will entity satisfy the filter if any of the values compares to the value specified in the filter in the manner described by the comparison operator.

so this should work will change!

Reply:
ok! If it doesn't, then we need to leave a TODO explaining the downsides, which we can talk about if necessary.